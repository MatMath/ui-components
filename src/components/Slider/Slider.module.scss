@use '../../colors/colors';
@use '../../typography/typography';

.sliderContainer {
  // CSS variables
  --label-color: #{colors.$dash-green-01};
  --label-disabled-color: #{colors.$dash-green-03};
  --thumb-size: 20px;
  --track-height: 4px;
  --track-fill-color: #{colors.$mid-green-00};
  --track-background-color: #{colors.$dash-green-05};
  --disabled-color: #{colors.$mid-green-03};
  --thumb-idle-color: #{colors.$mid-green-00};
  --thumb-hover-color: #{colors.$dash-green-00};
  --thumb-active-color: #{colors.$dash-darker-green-00};
  --tooltip-width: 42px;
  --tooltip-height: 36px;
  --tooltip-border-radius: 4px;
  --tooltip-background-color: #{colors.$dash-green-00};

  position: relative;
  display: grid;
  grid-template-columns: repeat(2, auto);
  grid-template-rows: 16px 20px 16px;
  grid-template-areas:
    ' label label '
    ' slider slider'
    ' min  max';

  .sliderLabel {
    grid-area: label;
    @include typography.label;
    margin-left: 4px;
    color: var(--label-color);
  }

  // Cross-browser compatible input range
  // https://css-tricks.com/styling-cross-browser-compatible-range-inputs-css/
  .sliderRange {
    @mixin track {
      cursor: pointer;
      height: var(--track-height);
      width: 100%;
    }

    @mixin thumb {
      background-color: var(--thumb-idle-color);
      border-radius: calc(var(--thumb-size) / 2);
      border: none;
      box-sizing: border-box;
      cursor: pointer;
      height: var(--thumb-size);
      width: var(--thumb-size);
    }

    /* Remove browser default style */
    -webkit-appearance: none; /* Hides the slider so that custom slider can be made */

    width: 100%; /* Specific width is required for Firefox. */
    height: var(--track-height);
    border-radius: 4px;
    grid-area: slider;
    align-self: center;
    // Fill percentage depends on value, evaluated in js
    --fill-percentage-value: 50%;
    background: linear-gradient(
      to right,
      var(--track-fill-color) var(--fill-percentage-value),
      var(--track-background-color) var(--fill-percentage-value)
    );

    /*
    * Removes the focus border.
    */
    &:focus {
      outline: none;
    }
    &::-moz-focus-outer {
      border: 0;
    }

    /* Thumb - the widget that you click or drag along the track */
    /* Special styling for WebKit/Blink */
    &::-webkit-slider-thumb {
      -webkit-appearance: none;
      @include thumb;
      margin-top: 0px; /* You need to specify a margin in Chrome, but in Firefox and IE it is automatic */
    }

    /* All the same for Firefox and IE */
    &::-moz-range-thumb {
      @include thumb;
    }
    &::-ms-thumb {
      @include thumb;
    }

    /* Track */
    &::-webkit-slider-runnable-track,
    &::-moz-range-track,
    &::-ms-track {
      @include track;
    }

    &:hover,
    &:focus {
      &::-webkit-slider-thumb {
        background-color: var(--thumb-hover-color);
      }
      &::-moz-range-thumb {
        background-color: var(--thumb-hover-color);
      }
      &::-ms-thumb {
        background-color: var(--thumb-hover-color);
      }
    }

    &:active {
      &::-webkit-slider-thumb {
        background: var(--thumb-active-color);
      }
      &::-moz-range-thumb {
        background: var(--thumb-active-color);
      }
      &::-ms-thumb {
        background: var(--thumb-active-color);
      }
      // Show tooltip on mouse pressed
      &:not(:disabled) + .tooltip {
        visibility: visible;
        opacity: 100;
      }
    }

    &:disabled {
      cursor: default;
      background: linear-gradient(
        to right,
        var(--disabled-color) var(--fill-percentage-value),
        var(--track-background-color) var(--fill-percentage-value)
      );

      &::-webkit-slider-thumb {
        cursor: default;
        background-color: var(--disabled-color);
      }
      &::-moz-range-thumb {
        cursor: default;
        background-color: var(--disabled-color);
      }
      &::-ms-thumb {
        cursor: default;
        background-color: var(--disabled-color);
      }

      & ~ .sliderMin {
        color: var(--label-disabled-color);
      }

      & ~ .sliderMax {
        color: var(--label-disabled-color);
      }
    }
  }

  /* Tooltip */
  .tooltip {
    visibility: hidden;
    opacity: 0;
    width: var(--tooltip-width);
    height: var(--tooltip-height);
    background-color: var(--tooltip-background-color);
    box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.2);
    border-radius: var(--tooltip-border-radius);
    position: absolute;
    top: -29px;
    text-align: center;
    transition: opacity 0.1s ease-in-out;

    .tooltipContent {
      display: inline-block;
      @include typography.paragraph-small;
      vertical-align: middle;
      color: colors.$white;
    }

    // Tooltip triangle
    // https://css-tricks.com/snippets/css/css-triangle/
    &::after {
      content: '';
      position: absolute;
      width: 0;
      height: 0;
      top: 100%;
      left: 50%;
      margin-left: -6px;
      border-left: 6px solid transparent;
      border-right: 6px solid transparent;

      border-top: 6px solid var(--tooltip-background-color);
    }
  }

  .sliderMin {
    grid-area: min;
    @include typography.body-small;
    color: var(--label-color);
    justify-self: start;
    margin-left: 8px;
  }

  .sliderMax {
    grid-area: max;
    @include typography.body-small;
    color: var(--label-color);
    justify-self: end;
    margin-right: 4px;
  }
}
