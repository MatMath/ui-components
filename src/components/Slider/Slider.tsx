import * as React from 'react';
import styles from './Slider.module.scss';
import { getClassNames } from '@utility/cssUtils';

export interface SliderProps {
  label: string;
  value: number;
  disabled?: boolean;
  max?: number;
  min?: number;
  onChange?(value: number): void;
  onChangeComplete?(value: number): void;
  step?: number;
  tooltip?: boolean;
  className?: string;
}

export const Slider = ({
  label,
  value,
  className = '',
  min = 0,
  max = 100,
  step = 1,
  tooltip = true,
  onChange,
  onChangeComplete,
  ...otherProps
}: SliderProps) => {
  const sliderRef = React.useRef<HTMLInputElement>(null);
  const tooltipRef = React.useRef<HTMLDivElement>(null);

  React.useEffect(() => {
    const slider = sliderRef.current;

    if (!slider) {
      return;
    }

    const range = max - min;
    const valuePercentage = ((value - min) / range) * 100;
    slider.style.setProperty('--fill-percentage-value', `${valuePercentage}%`);
  }, [sliderRef, value, max, min]);

  const handleChangeDone = () => {
    onChangeComplete && onChangeComplete(value);
  };

  React.useEffect(() => {
    const slider = sliderRef.current;

    if (!slider) {
      return;
    }

    // Add event listener
    slider.addEventListener('mouseup', handleChangeDone);
    slider.addEventListener('keyup', handleChangeDone);

    // Remove event listener on cleanup
    return () => {
      slider.removeEventListener('mouseup', handleChangeDone);
      slider.removeEventListener('keyup', handleChangeDone);
    };
  }, [sliderRef, value, handleChangeDone]);

  React.useEffect(() => {
    const { current: slider } = sliderRef;
    const { current: tooltipElement } = tooltipRef;
    if (!slider || !tooltipElement) {
      return;
    }

    const halfThumbWidth =
      parseInt(
        getComputedStyle(slider)
          .getPropertyValue('--thumb-size')
          .replace('px', '')
      ) / 2;

    const halfTooltipWidth = tooltipElement.clientWidth / 2 - 2;

    const centerPosition = slider.clientWidth / 2;

    const percentOfRange = (value - min) / (max - min);
    const valuePxPosition = percentOfRange * slider.clientWidth;
    const distFromCenter = valuePxPosition - centerPosition;
    const percentDistFromCenter = distFromCenter / centerPosition;

    const offset = percentDistFromCenter * halfThumbWidth;

    const tooltipPosition = valuePxPosition - halfTooltipWidth - offset;

    tooltipElement.style.left = `${tooltipPosition}px`;
  }, [value, sliderRef, tooltipRef, max, min]);

  return (
    <div className={getClassNames(styles.sliderContainer, className)}>
      <label htmlFor={`slider-${label}`} className={styles.sliderLabel}>
        {label} ({value})
      </label>
      <input
        {...otherProps}
        name={`slider-${label}`}
        type='range'
        ref={sliderRef}
        min={min}
        value={value}
        max={max}
        step={step}
        className={styles.sliderRange}
        onChange={event => {
          const floatValue = parseFloat(event.currentTarget.value);
          if (!isNaN(floatValue) && onChange) {
            onChange(floatValue);
          }
        }}
      />
      {tooltip ? (
        <output
          className={styles.tooltip}
          ref={tooltipRef}
          htmlFor={`slider-${label}`}
        >
          <span className={styles.tooltipContent}>{value}</span>
        </output>
      ) : null}
      <div className={styles.sliderMin}>{min}</div>
      <div className={styles.sliderMax}>{max}</div>
    </div>
  );
};
