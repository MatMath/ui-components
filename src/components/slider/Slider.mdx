---
name: Slider
menu: Components
---

import { Playground, Props } from 'docz';
import { Slider } from './Slider.tsx';

# Slider

A slider

## Properties

<Props of={Slider} />

## Usage

### Basic usage

Examples below show usage of the Slider

<Playground
  style={{
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    height: 430,
    flex: 1
  }}
>
  {() => {
    const [value, setValue] = React.useState(50);
    const [value2, setValue2] = React.useState(30);
    return (
      <>
        <Slider
          value={value}
          onChange={setValue}
          label='Slider'
          min={0}
          max={100}
          onChangeComplete={val => console.log(`done ${val}`)}
        />
        <Slider
          value={value2}
          onChange={setValue2}
          label='Disabled'
          min={0}
          max={100}
          disabled
          onChangeComplete={val => console.log(`done ${val}`)}
        />
      </>
    );
  }}
</Playground>
