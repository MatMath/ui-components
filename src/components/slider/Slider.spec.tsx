import * as React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { Slider } from './Slider';

describe('<Slider />', () => {
  it('should render default', () => {
    const label = 'Slider';
    const value = 50;
    const { asFragment } = render(<Slider value={value} label={label} />);

    expect(asFragment()).toMatchSnapshot();

    expect(screen.queryByText(label)).toBeInTheDocument();
    expect(screen.queryByText('0')).toBeInTheDocument();
    expect(screen.queryByText('100')).toBeInTheDocument();
  });

  it('should render with a tooltip on press', () => {
    const value = 30;
    const label = 'Slider';
    render(<Slider value={value} label={label} />);
    const slider = screen.getByLabelText(label);
    fireEvent.mouseDown(slider);
    expect(screen.queryByText(value.toString())).toBeInTheDocument();
  });

  it('should not render with a tooltip on press when tooltip is set to false', () => {
    const value = 30;
    const label = 'Slider';
    render(<Slider value={value} label={label} tooltip={false} />);
    const slider = screen.getByLabelText(label);
    fireEvent.mouseDown(slider);
    expect(screen.queryByText(value.toString())).not.toBeInTheDocument();
  });
});
