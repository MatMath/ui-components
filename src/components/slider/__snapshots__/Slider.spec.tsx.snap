// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider /> should render default 1`] = `
<DocumentFragment>
  <div
    class="sliderContainer"
  >
    <label
      class="sliderLabel"
      for="slider"
    >
      Slider
    </label>
    <input
      class="sliderRange"
      id="slider"
      max="100"
      min="0"
      step="1"
      type="range"
      value="50"
    />
    <output
      class="tooltip"
      for="slider"
    >
      <span
        class="tooltipContent"
      >
        50
      </span>
    </output>
    <div
      class="sliderMin"
    >
      0
    </div>
    <div
      class="sliderMax"
    >
      100
    </div>
  </div>
</DocumentFragment>
`;
